File: implement-shell-tools/cat/cat.js
const fs = require("fs");

const args = process.argv.slice(2);

const showLineNumbers = args.includes("-n");
const numberNonEmptyLines = args.includes("-b");

const files = args.filter(arg => arg !== "-n" && arg !== "-b");

for (const file of files) {
  const content = fs.readFileSync(file, "utf8");

  if (numberNonEmptyLines) {
    const lines = content.split("\n");
    let lineNumber = 1;

    lines.forEach((line) => {
      if (line.trim() !== "") {
        console.log(`${lineNumber} ${line}`);
        lineNumber++;
      } else {
        console.log(line);
      }
    });
  } else if (showLineNumbers) {
    const lines = content.split("\n");

    lines.forEach((line, index) => {
      console.log(`${index + 1} ${line}`);
    });
  } else {
    process.stdout.write(content);
  }
}

File: implement-shell-tools/cat/cat.py
import sys

args = sys.argv[1:]

show_line_numbers = "-n" in args
number_non_empty = "-b" in args

files = [arg for arg in args if arg not in ["-n", "-b"]]

for file in files:
    with open(file, "r") as f:
        lines = f.readlines()

        if number_non_empty:
            count = 1
            for line in lines:
                if line.strip() != "":
                    print(f"{count:6}  {line}", end="")
                    count += 1
                else:
                    print(line, end="")

        elif show_line_numbers:
            for i, line in enumerate(lines, start=1):
                print(f"{i:6}  {line}", end="")

        else:
            print("".join(lines), end="")